### ISSUE TITLE
Change 70b7ce6 breaks ability to manage and mount Grafana config from Docker host

### ISSUE THREAD

**Issue by dbeckham** [grafana/grafana-docker#61]
Change 70b7ce6a9dd7fc9baa8a5677ee184223cdbe3249 from Mar. 21st breaks the ability for a configuration management system (Puppet, Chef, etc.) to manage a bind mounted Grafana config, or grafana configuration directory. Whatever security and permissions that the host have set are overridden by the `run.sh` startup file.

A simple work around might be to convert the `grafana.ini` file to environment variables, but **you can not fully configure this Docker container using environment variables because LDAP support specifically requires an externally managed `*.toml` configuration file.**

Is it really necessary to change the permissions of every single file, even if the Grafana user can already read them?  It feels like bad security to go to the trouble of using gosu, but then force the system to allow the unprivileged grafana user to potentially write to sensitive files mounted from the host.

**Comment by jangaraj** [comment]
IMHO it's normal behavior, for example elasticsearch has the same problem.

Other possible workarounds:
- don't manage full config life cycle (or permissions), only initial content (Puppet `ensure=>present`)
- Dockerize your config files -> create container only with config files and "mount" them to the grafana container by using `--volume-from` = no writes to the host filesystem, but you will need to implement backups of config container

**Comment by jjaimon** [comment]
This issue breaks the convenience of running in a container.  It will be easier if there is a a way to specify the uid of grafana when the container is created and use that for setting the permissions.   For various production reasons, we'll need to maintain data and configurations in a persistent storage.

**Comment by danielsmith-eu** [comment]
Agree with @jjaimon above, this means you can't run grafana at all if the volumes are bound to the host filesystem, which is a usual use case of docker.

**Comment by kaihendry** [comment]
Currently I am losing my panel changes & suffering with "Invalid alert data. Cannot save dashboard", so I am here since I guess I need persistent storage.

Am I right in thinking I can't simply bind this image to a volume mount for persistent storage? I certainly do not want to run another image for persistent storage as asked for on https://github.com/grafana/grafana-docker#grafana-container-with-persistent-storage-recommended

Omg.. Docker is such a emotional roller coaster.

**Comment by torkelo** [comment]
Do you have a recommended change? The container user needs read permissions for configuration and read write permissions for data volume, very standard docker behavior to ensure those permissions on startup

**Comment by kaihendry** [comment]
Sorry, I am not familiar with the `volumes-from` concept. Isn't there no way to ensure correct permissions on a straightforward volume mount?

Myself and other newbs might be put off Grafana, since now we have to figure out how to maintain **grafana-storage**. My monitoring / graphing stack http://github.com/kaihendry/pingprom already is a huge burden on my tiny brain.

**Comment by torkelo** [comment]
> since now we have to figure out how to maintain grafana-storage

Not sure I follow. maintain what? and why?

**Comment by bergquist** [comment]
ref https://github.com/grafana/grafana-docker/issues/51

**Comment by sandersaares** [comment]
I suspect this is related to https://github.com/grafana/grafana/issues/9073

**Comment by sandersaares** [comment]
> very standard docker behavior to ensure those permissions on startup

I question this statement, as I have never seen a container that does this.

My expectations are:

1. Docker image is built with proper permissions already set up in case a "no external storage" setup is supported.
1. Is external storage is mapped into the container by the operator at deploy-time, it is the responsibility of the operator to ensure read/write rights are correctly set up. In my experience, Docker runs as root on Linux so there's not really much you need (but I admit this may vary by operating system? For sure there is more to it on Windows).

At no point do I expect anything inside the container to start touching the permission of the data volumes I provide to the container.

**Comment by infinityminglei** [comment]
I think we are experiencing the same issue with this when configuring the LDAP.
We need to configure LDAP for our grafana instances, so we are trying to manage ldap.toml configuration file on the host file system and add it as the docker volume. But we are getting permission denied issue like `grafana    | t=2017-08-28T15:57:39+0000 lvl=crit msg="Failed to load ldap config file" logger=ldap error="open /opt/grafana/config/ldap.toml: permission denied"`, we have to explicitly set `/opt/grafana/config/ldap.toml` permission as 601 to let grafana open it. Can you give me some suggestions for how to configure this toml file externally instead of change it to this weird file permission?

**Comment by sandersaares** [comment]
I eventually arrived at some sort of secure steady state where Grafana worked without the need for world-readable LDAP configuration. I paste the permission flags of my directory tree, for your reference.

The config is at `/grafana/config/ldap.toml`, with `/grafana/config/` mapped to `/etc/grafana/` in the container.

Permissions for the filesystem entries are:

```
drwxrwxr-x   5 root root  4096 Aug 28 11:32 grafana
drwxrwx---  2 syslog crontab 4096 Aug 28 11:32 config
-rw-rw---- 1 syslog crontab 1594 Aug 28 11:31 ldap.toml
```

I am not 100% confident in that "syslog" and "crontab" (makes no sense to me) but attempting to access the config as non-sudo fails, so I guess it works?

**Comment by dbeckham** [comment]
This can easily be fixed by **not** requiring that an `ldap.toml` config file exist. This turned into a permissions discussions, but the problem is that the Grafana container is only partially configurable using environment variables. Make it **fully** configurable using environment variables, including the LDAP configuration, and this problem just goes away.

tl;dr The LDAP configuration for this container is super broken and should be configurable through environment variables like every other Grafana configuration option.

**Comment by taasjord** [comment]
Does Grafana really need write permission to it's configuration files? IMHO it is better to exit with error message if it lacks permissions, rather than to chown to syslog/crontab.

**Comment by xlson** [comment]
Fixed in 291c81788c0a823ce228c50d14713691a64390e2
